rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,265 @@
+# RFC 6330 - 2021-04-07 - Converting one log event into multiple log events
+
+This RFC describes an approach for transforming one input log event into multiple log events.
+
+## Scope
+
+In:
+
+- Turning a log event into multiple log events
+
+Out:
+
+- Turning a log event into multiple metrics
+- Turning a metric into multiple metrics
+- Turning a metric into multiple logs
+
+## Motivation
+
+Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+## Internal Proposal
+
+The proposal is to:
+
+1. Extend `remap` so that if `.` is an array at the end, it will emit one event for each element in that array
+2. Add an `unnest` VRL function that will transform an object into an array of objects using a specified field on the input object
+3. Add an `unnest` transform that can be used for simple cases (could just "expand" into a `remap` transform)
```

Comment:
I would vote against a separate transform just for consistency and simplicity when browsing transforms.

Reply:
I tend to agree. I added it just based on earlier feedback on this proposal, but I'm inclined to leave it out. We can always add it later if the need arises.

Reply:
We can leave it off for now. I do agree if we go with the `.` solution the added transform is less of a requirement.

If we went with the previous proposal of using the final value of the program to determine the events triggered, then we'd need a new transform, since that's how how `remap` currently works.